server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -972,6 +972,16 @@ public Object[] addGeneratedValues(IndexItem item) {
         List<Object> extendedValues = new ArrayList<>(insertValues.length);
         Collections.addAll(extendedValues, insertValues);
 
+        // INSERT... ON CONFLICT or UPDATE statements might not contain some columns.
+        // Unused columns are anyway added to the targets by the UpdateToInsert since we need all columns to fetch conflicting values.
+        // However, insertValues contains only provided values.
+        // We explicitly add NULL-s for unused columns to avoid wrong binding of computed non-deterministic values to unused columns.
+        // Note, that those NULL-s won't be written to the source since Indexer ignores NULL-s.
+        int numOfUnusedColumns = columns.size() - extendedValues.size();
+        for (int i = 0; i < numOfUnusedColumns; i++) {
+            extendedValues.add(null);
+        }
```

Comment:
If generated column is provided, we don't validate it and take it as is (https://github.com/crate/crate/pull/14323)
Then expansion logic is not even hit

Reply:
Not sure whether we need this test as I also added IndexerTest.

1. I fell like `IndexerTest` doesn't cover all interactions between Indexer/UpdateToInsert/TransportShardUpsertAction. For example, unit test relies on specific behavior of another component when creating Indexer object
https://github.com/crate/crate/pull/15309/files#diff-5679257d0b0f40d3153d406ef4477bd35e7d8ee1e5c98f0c640c95afe3c2e041R986 which I feel like not really stable.
2. If _raw selection is too hacky, I can use random like in `test_adds_non_deterministic_sub_columns_when_root_is_not_in_targets` to avoid jdbc testing issue highlighted in https://github.com/crate/crate/pull/14323